packages/vm/src/instance.rs

```diff
@@ -84,102 +78,135 @@ where
-                // This is never written to chain.
-                // Takes a pointer argument of a memory region that must contain an UTF-8 encoded string.
-                // Ownership of both input and output pointer is not transferred to the host.
-                "debug" => Func::new(move |ctx: &mut Ctx, message_ptr: u32|-> VmResult<()> {
-                    if print_debug {
-                        print_debug_message(ctx, message_ptr)?;
-                    }
-                    Ok(())
-                }),
-                "query_chain" => Func::new(move |ctx: &mut Ctx, request_ptr: u32| -> VmResult<u32> {
-                    do_query_chain::<S, Q>(ctx, request_ptr)
-                }),
-            },
-        });
 
+        let store = module.store();
+
+        let env = Env::new(gas_limit);
+
+        let i32_to_void = FunctionType::new(vec![Type::I32], vec![]);
+        let i32i32_to_i32 = FunctionType::new(vec![Type::I32, Type::I32], vec![Type::I32]);
+
+        let mut import_obj = ImportObject::new();
+        let mut env_imports = Exports::new();
+
+        // Reads the database entry at the given key into the the value.
+        // Returns 0 if key does not exist and pointer to result region otherwise.
+        // Ownership of the key pointer is not transferred to the host.
+        // Ownership of the value pointer is transferred to the contract.
+        env_imports.insert(
+            "db_read",
+            Function::new_native_with_env(store, env.clone(), native_db_read),
+        );
+
+        // Writes the given value into the database entry at the given key.
+        // Ownership of both input and output pointer is not transferred to the host.
+        env_imports.insert(
+            "db_write",
+            Function::new_native_with_env(store, env.clone(), native_db_write),
+        );
+
+        // Removes the value at the given key. Different than writing &[] as future
+        // scans will not find this key.
+        // At the moment it is not possible to differentiate between a key that existed before and one that did not exist (https://github.com/CosmWasm/cosmwasm/issues/290).
+        // Ownership of both key pointer is not transferred to the host.
+        env_imports.insert(
+            "db_remove",
+            Function::new_native_with_env(store, env.clone(), native_db_remove),
+        );
+
+        // Reads human address from source_ptr and writes canonicalized representation to destination_ptr.
+        // A prepared and sufficiently large memory Region is expected at destination_ptr that points to pre-allocated memory.
+        // Returns 0 on success. Returns a non-zero memory location to a Region containing an UTF-8 encoded error string for invalid inputs.
+        // Ownership of both input and output pointer is not transferred to the host.
+        env_imports.insert(
+            "canonicalize_address",
+            Function::new_with_env(store, &i32i32_to_i32, env.clone(), move |env, args| {
```

Comment:
Ah, the main difference here is not performance per-say but needing to write this explicit wrapper, where `new_native_with_env` will auto-generate this wrapper and proper calling types. Correct?

So, if moving the Api around is more trouble than a few of these wrapper functions it is not worth it.

Reply:
The main difference is if we capture variables or not